lib/scheduler/thread_pool.rb

```diff
@@ -0,0 +1,138 @@
+# frozen_string_literal: true
+
+module Scheduler
+  # ThreadPool manages a pool of worker threads that process tasks from a queue.
+  # It maintains a minimum number of threads and can scale up to a maximum number
+  # when there's more work to be done.
+  class ThreadPool
+    class ShutdownError < StandardError
+    end
+
+    def initialize(min_threads:, max_threads:, idle_time:)
+      raise ArgumentError, "min_threads must be positive" if min_threads <= 0
+      raise ArgumentError, "max_threads must be >= min_threads" if max_threads < min_threads
+      raise ArgumentError, "idle_time must be positive" if idle_time <= 0
+
+      @min_threads = min_threads
+      @max_threads = max_threads
+      @idle_time = idle_time
+
+      @threads = []
+      @queue = Queue.new
+      @mutex = Mutex.new
+      @new_work = ConditionVariable.new
+      @shutdown = false
+
+      # Initialize minimum number of threads
+      @min_threads.times { spawn_thread }
+    end
+
+    def post(&block)
+      raise ShutdownError, "Cannot post work to a shutdown ThreadPool" if shutdown?
+
+      db = RailsMultisite::ConnectionManagement.current_db
+      wrapped_block = wrap_block(block, db)
+
+      @mutex.synchronize do
+        @queue << wrapped_block
+        @new_work.signal
+        spawn_thread if @threads.size < @max_threads
+      end
+    end
+
+    def shutdown(timeout: 30)
+      @mutex.synchronize do
+        @shutdown = true
+        @threads.size.times { @queue << :shutdown }
+        @new_work.broadcast
+      end
+
+      # Copy threads array to avoid concurrent modification
+      threads_to_join = nil
+      @mutex.synchronize { threads_to_join = @threads.dup }
+
+      failed_to_shutdown = false
+
+      deadline = Process.clock_gettime(Process::CLOCK_MONOTONIC) + timeout
+      threads_to_join.each do |thread|
+        remaining_time = deadline - Process.clock_gettime(Process::CLOCK_MONOTONIC)
+        break if remaining_time <= 0
+        if !thread.join(remaining_time)
+          Rails.logger.error "ThreadPool: Failed to join thread within timeout"
+          failed_to_shutdown = true
+        end
+      end
+
+      raise ShutdownError, "Failed to shutdown ThreadPool within timeout" if failed_to_shutdown
+    end
+
+    def shutdown?
+      @shutdown
```

Comment:
I think we need to access the instance variable within a mutex here.

Reply:
technically yeah, I guess will change.